design/design-doc-package-refactor.md

```diff
@@ -0,0 +1,690 @@
+templating controller have the additional step of first producing a
+`StackDefinition` in the install `Job`, then translating that to a `Package` in
+the [_`StackDefinition`
+controller_](https://github.com/crossplane/crossplane/blob/6fc50822fbf11a7d31f8a9dabde5c8948c3b36ac/pkg/controller/packages/templates/stackdefinition_controller.go#L53).
+
+### Future Workflow
+
+The first step towards supporting the upgrade of packages is by introducing the
+`PackageRevision` type. The `PackageRevision` type will look very similar to the
+current `Package`, but will add the following fields:
+
+- `spec.infrastructureDefinitions`: list of `InfrastructureDefinition`s
+  installed by the package.
+- `spec.compositions`: list of `Composition`s installed by the package.
+- `spec.state`: the desired state, which can be one of `Pending`, `Active`,
+  `Stopping` or `Inactive`.
+
+The following fields will be removed:
+
+- `spec.packageType`: currently is one of `Provider`, `Stack`, `Application`, or
+  `Addon`.
+- `spec.permissionType`: currently is one of `Cluster` or `Namespace`.
+
+In addition, `PackageRevision` will be a cluster-scoped type, rather than the
+current namespace-scoped `Package`. The current schema for `Package` can be
+viewed
+[here](https://doc.crds.dev/github.com/crossplane/crossplane/cluster/charts/crossplane-types/crds/packages.crossplane.io_packages.yaml@v0.12.0).
+
+In order to continue to support the current functionality of
+`ClusterPackageInstall`, `PackageInstall`, `StackDefinition` the underlying
+`Package` type will be renamed to `LegacyPackage`. The existing `Package` type
+will be refactored to look like the current `ClusterPackageInstall` schema with
+thw following modifications.
+
+This field will be added:
+
+- `status.currentHash`: a string that is the `sha256` hash of the image in the
+  current `PackageRevision`.
+- `status.currentImage`: a string that is the name of the current
+  `PackageRevision` image.
+
+The following fields will be removed:
+
+- `status.installJob`: a reference to the install `Job` for the
+  `ClusterPackageInstall`. 
+- `status.packageRecord`: a reference to the `Package` that represents the
+  installation unit for the `ClusterPackageInstall`. The `PackageRevision` is
+  now discoverable by name at the cluster scope.
+
+
+Before outlining the functionality of the new `Package` and `PackageRevision`
+controllers, it is useful to establish the responsibilities of each.
+
+The `Package` controller is responsible for:
+
+- Managing the lifecycle of its `PackageRevision`s
+- Managing the lifecycle of any objects installed by its `PackageRevision`s
+
+The `PackageRevision` controller is responsible for:
+
+- Unpacking a specific package image
+- Managing the lifecycle of any controller is installed by a package image
+- Managing the ownership of objects included in the package
+
+Note that the `Package` controller is responsible for resources that may persist
```

Comment:
if the PackageRevision controller unpacks the image, how does the Package controller create/delete/manage the CRDs? where does it get them from?

Reply:
Need an update here to reflect https://github.com/crossplane/crossplane/pull/1616#discussion_r448447207. However, if we did want the `PackageRevision` to unpack itself (which would cause complications with the current model of naming it after its hash, see https://github.com/crossplane/crossplane/pull/1616#discussion_r448359703) it would be possible for the `Package` controller to still get this `Job` by following the `PackageRevision` to it. As mentioned though, this would be a pretty big change from current model.